### ISSUE TITLE
Using the github <-> Zenodo integration mutually exclusive with zenodraft action?

### ISSUE THREAD

**Issue by alexlancaster** [zenodraft/action#22]
After getting duplicate uploads when enabling both the zenodraft action and the integration github <-> zenodo here: https://sandbox.zenodo.org/account/settings/github/  I finally realized that (I think) they are mutually exclusive, is that correct?

I am getting a separate upload from zenodraft, and it also doesn't seem to include the version taken from `git tag` as the integration does.  Has the integration superceded this action?  One thing that the integration doesn't do, which this action does, it up date the `CITATION.cff` file, which is the reason I tried using it, but it loses the ability to use the git tag.

**Comment by jspaaks** [comment]
Hi @alexlancaster, thanks for making the issue. Indeed the GitHub integration and `zenodraft` should be mutually exclusive. The GitHub integration predates `zenodraft` by a few years.

> it also doesn't seem to include the version taken from `git tag`

Correct, it tries to use metadata bundled with the repo files, e.g. from `version` in the file listed as `metadata`'s value.

PS which repo does this happen on? It's always useful for me to see how `zenodraft` is being used "in the wild", so to speak.

**Comment by alexlancaster** [comment]
> Hi @alexlancaster, thanks for making the issue. Indeed the GitHub integration and `zenodraft` should be mutually exclusive. The GitHub integration predates `zenodraft` by a few years.
> 
> > it also doesn't seem to include the version taken from `git tag`
> 
> Correct, it tries to use metadata bundled with the repo files, e.g. from `version` in the file listed as `metadata`'s value.

Is there a way to handle it like the integration does? I asked because I set the version by creating a tag with the version number during the release and that's the only place that I ever set the version number. The integration nicely passes this along to zenodo.

> PS which repo does this happen on? It's always useful for me to see how `zenodraft` is being used "in the wild", so to speak.
> 

Sure, it's in the pypop repo in a branch right now:

https://github.com/alexlancaster/pypop/tree/zenodo_upload
 
https://github.com/alexlancaster/pypop/blob/00fb19161d053fef30f26a28a744fca9674e1fce/.github/workflows/build_wheels.yml#L123

PS. It seems that as of a couple of days ago, sandbox.zenodo.org isn't working properly with this migration going on, so I haven't been able to test it for a few days. Have you had the same problem?

**Comment by jspaaks** [comment]
OK, so if I understand you correctly you want the tag that is entered in GitHub when you make the release to be the single source of truth for the version string. 

How about something like this using `jq` and `yq` (command line programs to manipulate JSON and YAML respectively):

```yaml
jobs:
  publish_zenodo:
    name: Publish to zenodo on production releases only
    runs-on: ubuntu-latest
    if: github.event_name == 'release' && github.event.action == 'published' && '!github.event.release.prerelease'
    steps:
       - name: Checkout the contents of your repository
         uses: actions/checkout@v4
       - name: Upsert the version in the metadata in .zenodo.json
         run: |
           cp .zenodo.json .zenodo.orig.json
           echo "{\"version\": \"$GITHUB_REF_NAME\"}" | cat .zenodo.orig.json - | jq -s add > .zenodo.json
           rm .zenodo.orig.json
       - name: Upsert the version in the metadata in CITATION.cff
         run: |
           yq -i ".version = \"$GITHUB_REF_NAME\"" CITATION.cff
       - name: Push changes back to repo files
         run: |
           git config --global user.name ci
           git config --global user.email "ci@users.noreply.github.com"
           git add .zenodo.json CITATION.cff
           git commit -m "Upserted version into CITATION.cff and into .zenodo.json"
           git push origin $GITHUB_REF_NAME
       - name: Create a draft snapshot of your repository contents as a new
               version in collection 1243488 on Zenodo Sandbox using metadata
               from repository file .zenodo.json
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           ZENODO_SANDBOX_ACCESS_TOKEN: ${{ secrets.ZENODO_SANDBOX_ACCESS_TOKEN }}
         uses: zenodraft/action@0.10.0
         with:
           collection: 1243488
           metadata: .zenodo.json
           publish: true
           sandbox: true
           compression: tar.gz
           upsert-doi: true
           upsert-location: doi
```

BTW I just tried using Zenodo Sandbox and indeed that seems to have changed in the last few days, probably something to do with the migration. I don't know if zenodraft will keep working once the migration is done, there could be API changes maybe. So anyway I tested the above on one of my collections on real Zenodo, that seems to work like it used to but I expect it will change once the Zenodo folks like what they are seeing with Sandbox.

O BTW I had to update your `.zenodo.json`, it had a trailing comma near the bottom which `jq` didn't like. Also, there was a version string in CITATION.cff that gets overridden with the value from the tag now. Just so you know.

PS my fork of the workflow with my changes is here for the time being https://github.com/jspaaks/pypop/blob/32fa216ef7bbbbfd17319ad80cbc50f4c43afbd4/.github/workflows/build_wheels.yml#L1

Hope this helps you! Best regards,
-Jurriaan

**Comment by alexlancaster** [comment]
> OK, so if I understand you correctly you want the tag that is entered in GitHub when you make the release to be the single source of truth for the version string.

yes, exactly!

> How about something like this using `jq` and `yq` (command line programs to manipulate JSON and YAML respectively):
> 
> ```yaml
> jobs:
>   publish_zenodo:
>     name: Publish to zenodo on production releases only
>     runs-on: ubuntu-latest
>     if: github.event_name == 'release' && github.event.action == 'published' && '!github.event.release.prerelease'
>     steps:
>        - name: Checkout the contents of your repository
>          uses: actions/checkout@v4
>        - name: Upsert the version in the metadata in .zenodo.json
>          run: |
>            cp .zenodo.json .zenodo.orig.json
>            echo "{\"version\": \"$GITHUB_REF_NAME\"}" | cat .zenodo.orig.json - | jq -s add > .zenodo.json
>            rm .zenodo.orig.json
>        - name: Upsert the version in the metadata in CITATION.cff
>          run: |
>            yq -i ".version = \"$GITHUB_REF_NAME\"" CITATION.cff
>        - name: Push changes back to repo files
>          run: |
>            git config --global user.name ci
>            git config --global user.email "ci@users.noreply.github.com"
>            git add .zenodo.json CITATION.cff
>            git commit -m "Upserted version into CITATION.cff and into .zenodo.json"
>            git push origin $GITHUB_REF_NAME
>        - name: Create a draft snapshot of your repository contents as a new
>                version in collection 1243488 on Zenodo Sandbox using metadata
>                from repository file .zenodo.json
>          env:
>            GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
>            ZENODO_SANDBOX_ACCESS_TOKEN: ${{ secrets.ZENODO_SANDBOX_ACCESS_TOKEN }}
>          uses: zenodraft/action@0.10.0
>          with:
>            collection: 1243488
>            metadata: .zenodo.json
>            publish: true
>            sandbox: true
>            compression: tar.gz
>            upsert-doi: true
>            upsert-location: doi
> ```

That is awesome. I will try and create a PR that merges this into my `zenodo_upload` branch.  Just so I understand it, it looks like it actually gets the prereserved DOI from Zenodo, and then updates the `.cff` and `.zenodo.json` files before retagging with the github tag?  And then the tar/zipball that is generated for the upload will include all the files in that tag as well?  If so, that is very nice.  So the tagged version includes the correct metadata in the tarball.  That would seem to solve the chicken-or-egg problem I was having with having either the tag ahead or behind of the tarball on zenodo before.

Did it actually upload to the main zenodo, is there test server or place where I can see the final uploaded version, since the sandbox, as you note below, isn't working for you either.

> BTW I just tried using Zenodo Sandbox and indeed that seems to have changed in the last few days, probably something to do with the migration. I don't know if zenodraft will keep working once the migration is done, there could be API changes maybe. So anyway I tested the above on one of my collections on real Zenodo, that seems to work like it used to but I expect it will change once the Zenodo folks like what they are seeing with Sandbox.

Yeah, according the blog announcement all existing APIs should (theoretically) continue to work for up a year to give people time to adjust, so I would expect that sandbox should continue to work, although for the brief time I was testing I did see some differences (e.g. it didn't respect the version string).  Those problems might be the reason they postponed the merge until mid-October.


> O BTW I had to update your `.zenodo.json`, it had a trailing comma near the bottom which `jq` didn't like. Also, there was a version string in CITATION.cff that gets overridden with the value from the tag now. Just so you know.

Got it. Good to know!

> PS my fork of the workflow with my changes is here for the time being https://github.com/jspaaks/pypop
/blob/32fa216ef7bbbbfd17319ad80cbc50f4c43afbd4/.github/workflows/build_wheels.yml#L1

Excellent.

> Hope this helps you! Best regards, -Jurriaan

It does!  Although just to close the loop on the original issue, it might be useful to add a big notice at the top of https://github.com/zenodraft/action that says something along the lines of "please disable the GitHub<->Zenodo integration before using this action".  I suspect many users will probably enable the integration by default when they first use Zenodo and either forget they also have it enabled, or don't realize that they are mutually exclusive, like me.  (I thought that they somehow worked together, initially).

Thanks again for your all your work on this action.  Maybe some of the tricks you supplied can make their way into the action itself eventually?

**Comment by jspaaks** [comment]
> That is awesome. I will try and create a PR that merges this into my zenodo_upload branch. Just so I understand it, it looks like it actually gets the prereserved DOI from Zenodo, and then updates the .cff and .zenodo.json files before retagging with the github tag? And then the tar/zipball that is generated for the upload will include all the files in that tag as well? If so, that is very nice. So the tagged version includes the correct metadata in the tarball. That would seem to solve the chicken-or-egg problem I was having with having either the tag ahead or behind of the tarball on zenodo before.

Thanks! The effects are like you describe, although getting the prereserved DOI happens after the .zenodo.json and CITATION.cff files are updated with the version from the tag ($GITHUB_REF_NAME), committed and pushed back to the repo. Basically, the workflow is aware of the version from the tag ($GITHUB_REF_NAME), it then puts that version in the relevant files, commits and pushes them, then asks Zenodo to reserve a DOI, then uses the DOI to update the CITATION.cff, again committing the changes to the repo, creates the tar and uploads it, then uses the metadata from .zenodo.json including the version from the tag to annotate the draft deposition.

> Did it actually upload to the main zenodo, is there test server or place where I can see the final uploaded version, since the sandbox, as you note below, isn't working for you either.

I first tried using the Sandbox, but (1) their site looked different and (2) it appeared to have problems, both of which I attributed to the migration work that was going on. So I opted for using the regular Zenodo instance (that looked like normal) by having `sandbox: false` and different credentials ($ZENODO_ACCESS_TOKEN instead of $ZENODO_SANDBOX_ACCESS_TOKEN). Because I was just doing some testing, I set `publish: false` and I also used a new collection by using the Zenodo form (Upload button > New upload > Add a nonempty file > Save but not publish the deposition will give you a new draft record number. Decrement by 1 to get the collection/concept number). It's not strictly necessary but this way it's easier to avoid clutter. So with that setup you get depositions on Zenodo, but they remain unfinalized until you click the button in Zenodo's form. You can see which unfinalized depositions you have by logging in to Zenodo and then clicking the "Upload" button near the top of the page, that will have the overview.

> Yeah, according the blog announcement all existing APIs should (theoretically) continue to work for up a year to give people time to adjust, so I would expect that sandbox should continue to work, although for the brief time I was testing I did see some differences (e.g. it didn't respect the version string). Those problems might be the reason they postponed the merge until mid-October.

Good to know. Migrating these kinds of systems is an absolute pain, I don't envy them. Looks like they're going about it in the right way, taking their time when needed, using the test env first, etc.

> it might be useful to add a big notice at the top of https://github.com/zenodraft/action that says something along the lines of "please disable the GitHub<->Zenodo integration before using this action"

Agreed, I'll leave this issue open. It would be good to have some message about that.

> Maybe some of the tricks you supplied can make their way into the action itself eventually?

Yes I'd like that. So far I haven't been successful in finding some time to work on this though, partly because there don't seem to be as many users as I anticipated which makes prioritizing this work over other work more difficult (a different kind of chicken and egg problem now that I think about it :). We'll see what happens after Zenodo finishes their migration I suppose, see if zenodraft is still useful afterwards.

**Comment by alexlancaster** [comment]
> I first tried using the Sandbox, but (1) their site looked different and (2) it appeared to have problems, both of which I attributed to the migration work that was going on. So I opted for using the regular Zenodo instance (that looked like normal) by having `sandbox: false` and different credentials ($ZENODO_ACCESS_TOKEN instead of $ZENODO_SANDBOX_ACCESS_TOKEN). Because I was just doing some testing, I set `publish: false` and I also used a new collection by using the Zenodo form (Upload button > New upload > Add a nonempty file > Save but not publish the deposition will give you a new draft record number. Decrement by 1 to get the collection/concept number). It's not strictly necessary but this way it's easier to avoid clutter. So with that setup you get depositions on Zenodo, but they remain unfinalized until you click the button in Zenodo's form. You can see which unfinalized depositions you have by logging in to Zenodo and then clicking the "Upload" button near the top of the page, that will have the overview.

@jspaaks I heard directly back from Zenodo that they have re-enabled the "old" sandbox again at sandbox.zenodo.org.  In the meantime have managed to the rest of this working (excluding the final deposit itself), and you can that I documented my progress on my PR starting at this comment https://github.com/alexlancaster/pypop/pull/129#issuecomment-1739808038

So I thought I'd try going back to using the sandbox again.  I created a new ` ZENODO_SANDBOX_ACCESS_TOKEN`, removed the `collection` ID temporarily (so it would create a fresh collection), and then re-enabled `zenodraft` in the workflow, but it's still failing at the final deposition stage:

https://github.com/alexlancaster/pypop/actions/runs/6387809657/job/17336666711#step:6:233

```
   Error: Something went wrong on POST to https://sandbox.zenodo.org/api/deposit/depositions: 403 - FORBIDDEN 
```
so I'm still stumped here.  I created a new `ZENODO_SANDBOX_ACCESS_TOKEN on Zenodo and added it to the repo but I'm not sure if this error on the `zenodraft` or the sandbox side.

> > it might be useful to add a big notice at the top of https://github.com/zenodraft/action that says something along the lines of "please disable the GitHub<->Zenodo integration before using this action"
> 
> Agreed, I'll leave this issue open. It would be good to have some message about that.

I can make a pull request to implement this if you like.

> > Maybe some of the tricks you supplied can make their way into the action itself eventually?
> 
> Yes I'd like that. So far I haven't been successful in finding some time to work on this though, partly because there don't seem to be as many users as I anticipated which makes prioritizing this work over other work more difficult (a different kind of chicken and egg problem now that I think about it :). We'll see what happens after Zenodo finishes their migration I suppose, see if zenodraft is still useful afterwards.

hopefully it will be!  meanwhile, I'm still trying to get the current sandbox working.

**Comment by alexlancaster** [comment]
> So I thought I'd try going back to using the sandbox again. I created a new `ZENODO_SANDBOX_ACCESS_TOKEN`, removed the `collection` ID temporarily (so it would create a fresh collection), and then re-enabled `zenodraft` in the workflow, but it's still failing at the final deposition stage:
> 
> https://github.com/alexlancaster/pypop/actions/runs/6387809657/job/17336666711#step:6:233
> 
> ```
>    Error: Something went wrong on POST to https://sandbox.zenodo.org/api/deposit/depositions: 403 - FORBIDDEN 
> ```
> 
> so I'm still stumped here. I created a new `ZENODO_SANDBOX_ACCESS_TOKEN` on the Zenodo and added it to the repo secreats, but I'm not sure if this error on the `zenodraft` or the sandbox side.
> 
> > > it might be useful to add a big notice at the top of https://github.com/zenodraft/action that says something along the lines of "please disable the GitHub<->Zenodo integration before using this action"

Answering my own question again - for some reason the token permission wasn't set right - so I enabled all 3 possible perms, and suddenly it works.  So now I have and end-to-end tested Zenodo upload: 

* https://sandbox.zenodo.org/record/1244745

and associated (updated) release: 

* https://github.com/alexlancaster/pypop/releases/tag/v0.9.75

All using release tag generated by the GitHub UI.  Finally!

One final niggle is that I've already setup `.gitattributes` to limit what goes in the default zip/tarballs generated by github itself (which work well - and which are also respected when using the integration method of Zenodo deposition), but apparently `zenodraft` doesn't take this into account.  I do this because there's a whole bunch of files that are really for the website (in the `website` directory) that shouldn't end up in the deposition.  Is there any easy way to restrict what goes in the tarballs?

**Comment by jspaaks** [comment]
Hey good job figuring all of that out! That's quite the workflow file you got there. And the deposition looks great. If you want even more metadata, maybe this will interest you: https://github.com/zenodraft/metadata-schema-zenodo

As to your final niggles: what if you add a step after checkout that runs `rm -r website/`? To take that one step further, I bet there is a way to take the list of files and the patterns from gitattributes, and produce a list of matching file names. You could then feed these into `rm` and have the list of relevant files that should be part of the deposition. Maybe https://stackoverflow.com/questions/29694540/list-all-files-having-a-git-attribute-set could be a starting point? Sidenote: these changes should not be committed back into your repo. 

Alternatively, if everything under `website/` is separate from the rest, you could split the repo. I do this as follows: 

1. make a new completely empty repo, say `pypop-website` using the github interface
2. on my local machine copy the whole directory `pypop` to a new directory `pypop-website`
3. `cd pypop-website`
4. now update git's `remote` as follows
5. `git remote remove origin`
6. `git remote add origin https://github.com/alexlancaster/pypop-website.git`
7. `git push origin main`
8. now https://github.com/alexlancaster/pypop-website should have an exact copy of the files, including git history. You can then delete parts in each of the the two repos as required.

I spawned an issue to update the text about tokens (#23).

I spawned an issue to add a note about the action conflicting with the GitHub-Zenodo integration (#24)

**Comment by alexlancaster** [comment]
> Hey good job figuring all of that out! That's quite the workflow file you got there. And the deposition looks great. If you want even more metadata, maybe this will interest you: https://github.com/zenodraft/metadata-schema-zenodo

Ah great!  I will look into that - definitely will want to set the grant metadata.  I couldn't see any easy way to do that via the integration - so it's good that I can set that here for all releases. (Zenodo doesn't seem to have any notion of 'global' metadata that is common for all items in collection - am I correct?  This is similar to PyPI where each upload is treated separately in terms of metadata, even when the releases are aggregated under a single name).

> As to your final niggles: what if you add a step after checkout that runs `rm -r website/`? To take that one step further, I bet there is a way to take the list of files and the patterns from gitattributes, and produce a list of matching file names. You could then feed these into `rm` and have the list of relevant files that should be part of the deposition. Maybe https://stackoverflow.com/questions/29694540/list-all-files-having-a-git-attribute-set could be a starting point? Sidenote: these changes should not be committed back into your repo.

Ah - yes - of course the github checkout persists to the next step in the workflow, so using some `git` magic as you suggested on the `.gitattributes` file (more robust than hardcoding the files to remove),  I came up with the following:

https://github.com/alexlancaster/pypop/blob/3e6578316a4f3182e52c6081ee199223325b1971/.github/workflows/build_wheels.yml#L73

This works, however because it only operates on files, it leaves a bunch of orphan directories in the final zipball, so that's the final piece I'm working on.  Suggestions welcome.

> Alternatively, if everything under `website/` is separate from the rest, you could split the repo. I do this as follows:

I thought about having a separate `website` repo initially, but the website build needs access to the source code to build the API and/or embed info about the command-line option in the `.rst` file, so I didn't go for that solution

> I spawned an issue to update the text about tokens (#23).
> 
> I spawned an issue to add a note about the action conflicting with the GitHub-Zenodo integration (#24)

Thanks!  I will close this issue then.  Perhaps I should convert it to a "discussion"?  Do you have that enabled for your repo?

**Comment by alexlancaster** [comment]
Closing. Original issue is now a new issue: #24

**Comment by alexlancaster** [comment]
> Ah - yes - of course the github checkout persists to the next step in the workflow, so using some `git` magic as you suggested on the `.gitattributes` file (more robust than hardcoding the files to remove), I came up with the following:
> 
> https://github.com/alexlancaster/pypop/blob/3e6578316a4f3182e52c6081ee199223325b1971/.github/workflows/build_wheels.yml#L73
> 
> ```
> git ls-files -z|git check-attr --stdin -z export-ignore |sed -zne 'x;n;n;s/^set$//;t print;b;:print;x;p'|xargs --null rm
> ```
> 
> This works, however because it only operates on files, it leaves a bunch of orphan directories in the final zipball, so that's the final piece I'm working on. Suggestions welcome.

I fixed this last one with a somewhat crude solution, after deleting the files, running:

```
find . -empty -type d -delete
```
Seems to work.

**Comment by jspaaks** [comment]
> (Zenodo doesn't seem to have any notion of 'global' metadata that is common for all items in collection - am I correct?

That is my understanding as well

>`git ls-files -z|git check-attr --stdin -z export-ignore |sed -zne 'x;n;n;s/^set$//;t print;b;:print;x;p'|xargs --null rm`

:muscle: 

> I thought about having a separate website repo initially, but the website build needs access to the source code to build the API and/or embed info about the command-line option in the .rst file, so I didn't go for that solution

makes sense

> Perhaps I should convert it to a "discussion"? Do you have that enabled for your repo?

I typically disable that, because I don't like having to check two places for every repo

**Comment by alexlancaster** [comment]
@jspaaks now that I've integrated your patch into pypop, I took the liberty of adding you to https://github.com/alexlancaster/pypop/blob/main/AUTHORS.rst along with your ORCID.  As per our contribution guidelines: http://pypop.org/docs/guide-chapter-contributing.html#crediting-contributors all code contributors are also eligible for inclusion on our  Zenodo release - bringing it full circle!  Let me know if you'd like to opt-out of any inclusion. Or if any of the affiliation data is incorrect.
